simulator_interface.cpp

```diff
@@ -0,0 +1,381 @@
+#endif
+
+using boost::asio::ip::tcp;
+using util::BinaryBuffer;
+
+typedef std::shared_ptr<BinaryBuffer> BinaryBufferPtr;
+
+/***************************** SimulatorInterface *****************************/
+SimulatorInterface* create_simulator(
+        const std::string& name, const int port_no) {
+
+    SimulatorInterface* g = NULL;
+
+    if (port_no >= 0) {
+        g = new SimulatorServer(name, port_no);
+    } else {
+        TeacherPtr teacher = nullptr;
+        SimulatorPtr game = nullptr;
+        if (name == "simple_game") {
+            game = std::make_shared<SimpleGame>();
+        } else if (name == "simple_race") {
+            game = std::make_shared<SimpleRaceGame>();
+        } else if (name == "xworld") {
+            FLAGS_color = true;
+
+            if (FLAGS_task_mode == "arxiv_lang_acquisition") {
+                FLAGS_task_groups_exclusive = false;
+            }
+
+            auto xwd = std::make_shared<XWorldSimulator>(true /*print*/);
+
+            int agent_id = xwd->add_agent();
+            game = std::make_shared<AgentSpecificSimulator>(xwd, agent_id);
+            teacher = std::make_shared<Teacher>(
+                    xwd->conf_file(), xwd, false /*print*/);
+        } 
+#ifdef ATARI
+        else if (name == "atari") {
+            game.reset(ArcadeGame::create());
+        }
+#endif
+        else {
+            LOG(FATAL) << "Unrecognized game type: " << name;
+        }
+        g = new SingleProcInterface(game, teacher);
+    }
+
+    return g;
+}
+
+/***************************** SingleProcInterface ****************************/
+SingleProcInterface::SingleProcInterface(SimulatorPtr game, TeacherPtr teacher)
+    : game_(game), teacher_(teacher) {}
+
+void SingleProcInterface::reset_game() {
+    game_->reset_game();
+    if (teacher_) {
+        teacher_->reset_after_game_reset();
+        teacher_->teach();
+    }
+}
+
+std::string SingleProcInterface::game_over_string() {
+    return GameSimulator::decode_game_over_code(game_->game_over());
+}
+
+int SingleProcInterface::game_over() {
+    return game_->game_over();
+}
+
+int SingleProcInterface::get_num_actions() { return game_->get_num_actions(); }
+
+int SingleProcInterface::get_lives() { return game_->get_lives(); }
+
+int64_t SingleProcInterface::get_num_steps() { return game_->get_num_steps(); }
+
+void SingleProcInterface::get_screen_out_dimensions(
+        size_t& height, size_t& width, size_t& channels) {
+    game_->get_screen_out_dimensions(height, width, channels);
+}
+
+void SingleProcInterface::show_screen(float reward) {
+    game_->show_screen(reward);
+}
+
+float SingleProcInterface::take_actions(const StatePacket& actions, int act_rep) {
+    float r = 0;
+    r += game_->take_actions(actions, act_rep);
+    if (teacher_) {
+        teacher_->teach();  // teacher reacts to agent's action and evaluate the
+                            // reward
+        r += teacher_->give_reward();
+    }
+
+    return r;
+}
+
```

Comment:
Add one more function call report_task_performance() which calls teacher_->report_task_performance() so that we don't need to expose teacher to game_player.